lib/auth/webauthnwin/webauthn_windows.go

```diff
@@ -0,0 +1,609 @@
+
+	modUser32               = windows.NewLazySystemDLL("user32.dll")
+	procGetForegroundWindow = modUser32.NewProc("GetForegroundWindow")
+)
+
+// nativeImpl keeps diagnostic informations about windows webauthn support.
+type nativeImpl struct {
+	webauthnAPIVersion int
+	hasCompileSupport  bool
+	isAvailable        bool
+	hasPlatformUV      bool
+}
+
+// newNativeImpl creates nativeImpl which contains diagnostics info.
+// Diagnostics are safe to cache because dll isn't something that
+// could change during program invocation.
+// Client will be always created, even if dll is missing on system.
+func newNativeImpl() *nativeImpl {
+	v, err := checkIfDLLExistsAndGetAPIVersionNumber()
+	if err != nil {
+		log.WithError(err).Warn("WebAuthnWin: failed to check version")
+		return &nativeImpl{
+			hasCompileSupport: true,
+			isAvailable:       false,
+		}
+	}
+	uvPlatform, err := isUVPlatformAuthenticatorAvailable()
+	if err != nil {
+		// This should not happen if dll exists, however we are fine with
+		// to proceed without uvPlatform.
+		log.WithError(err).Warn("WebAuthnWin: failed to check isUVPlatformAuthenticatorAvailable")
+	}
+
+	return &nativeImpl{
+		webauthnAPIVersion: v,
+		hasCompileSupport:  true,
+		hasPlatformUV:      uvPlatform,
+		isAvailable:        v > 0,
+	}
+}
+
+func (n *nativeImpl) CheckSupport() CheckSupportResult {
+	return CheckSupportResult{
+		HasCompileSupport:  n.hasCompileSupport,
+		IsAvailable:        n.isAvailable,
+		HasPlatformUV:      n.hasPlatformUV,
+		WebAuthnAPIVersion: n.webauthnAPIVersion,
+	}
+}
+
+// GetAssertion calls WebAuthNAuthenticatorGetAssertion endpoint from
+// webauthn.dll and returns CredentialAssertionResponse.
+// It interacts with both FIDO2 and Windows Hello depending on
+// loginOpts.AuthenticatorAttachment (using auto results in possibilty to select
+// either security key or Windows Hello).
+// It does not accept username - during passwordless login webauthn.dll provides
+// its own dialog with credentials selection.
+func (n *nativeImpl) GetAssertion(origin string, in *wanlib.CredentialAssertion, loginOpts *LoginOpts) (*wanlib.CredentialAssertionResponse, error) {
+	hwnd, err := getForegroundWindow()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	rpid, err := windows.UTF16PtrFromString(in.Response.RelyingPartyID)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	cd, jsonEncodedCD, err := clientDataToCType(in.Response.Challenge.String(), origin, string(protocol.AssertCeremony))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	opts, err := n.assertOptionsToCType(in.Response, loginOpts)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	var out *webauthnAssertion
+	ret, _, err := procWebAuthNAuthenticatorGetAssertion.Call(
+		uintptr(hwnd),
+		uintptr(unsafe.Pointer(rpid)),
+		uintptr(unsafe.Pointer(cd)),
+		uintptr(unsafe.Pointer(opts)),
+		uintptr(unsafe.Pointer(&out)),
+	)
+	if err != syscall.Errno(0) {
```

Comment:
Are you sure that the order here is correct? https://pkg.go.dev/golang.org/x/sys/windows#LazyProc.Call

> Callers must inspect the primary return value to decide whether an error occurred (according to the semantics of the specific function being called) before consulting the error.

From what I understand, `ret` should be checked before `err` here in all other cases.

Reply:
@jakule thanks for comment, i have added better handling of it in https://github.com/gravitational/teleport/pull/16634/commits/48e8b23ea71334935e84705e92379450e78b9cab